File: desktop/src/api/contratistas.ts
import { invoke } from "@tauri-apps/api/core";
import { solicitarSincronizacionNube } from "../eventosNube";

// Espejo de comandos/contratistas.rs y dto.rs (los tipos de filtro/edición
// son los DTO de frontera, no el FiltroContratistas/DatosActualizacionContratista
// reales del núcleo — esos tienen tipos como Igualdad<T> que no tiene sentido
// exponerle al webview tal cual).

export type TipoIngreso = "Praind" | "InHouse" | "PorCorreo" | "Swat";
export const TIPOS_INGRESO: TipoIngreso[] = ["Praind", "InHouse", "PorCorreo", "Swat"];

/** Texto de un tipo de ingreso en las grillas: en mayúsculas y con las
 * palabras separadas ("IN HOUSE", no el nombre interno "InHouse") --
 * pedido del usuario 2026-09-23. Mismo texto que el Excel/PDF del núcleo
 * (`historial::exportacion::tipo_texto`). `null` (fila remota vieja) → "—". */
export function textoTipoIngreso(tipo: TipoIngreso | null): string {
  switch (tipo) {
    case "Praind":
      return "PRAIND";
    case "InHouse":
      return "IN HOUSE";
    case "PorCorreo":
      return "POR CORREO";
    case "Swat":
      return "SWAT";
    default:
      return "—";
  }
}

export interface ContratistaResumen {
  id: number;
  empresa_id: number;
  cedula: string;
  nombre: string;
  empresa_nombre: string;
  tipo_ingreso: TipoIngreso;
  fecha_vencimiento_praind: string | null;
  es_personal_ruta: boolean;
  tiene_acceso: boolean;
  tiene_ingreso_activo: boolean;
}

export interface PaginaContratistas {
  items: ContratistaResumen[];
  total: number;
}

// Sólo texto: la grilla de Contratistas ya no lo usa (carga el universo
// completo y filtra/ordena del lado del cliente con AG Grid — ver
// Contratistas.tsx), pero el buscador en vivo de NuevoIngresoModal y
// GestionGafeteModal sigue necesitando una búsqueda de texto contra el
// servidor.
export interface FiltroContratistas {
  texto?: string;
}

export interface DatosContratista {
  cedula: string;
  nombre: string;
  empresa_id: number;
  tipo_ingreso: TipoIngreso;
  fecha_vencimiento_praind: string | null;
  es_personal_ruta: boolean;
  tiene_acceso: boolean;
}

/// Espejo de Contratista::requiere_praind() (src/models/contratista.rs) — el
/// core es quien manda esta regla, acá sólo se replica para decidir cuándo
/// mostrar/exigir el campo de fecha en el formulario. La validación real
/// sigue pasando por el backend de todos modos.
export function requierePraind(datos: {
  es_personal_ruta: boolean;
  tipo_ingreso: TipoIngreso;
}): boolean {
  return (
    datos.es_personal_ruta ||
    datos.tipo_ingreso === "Praind" ||
    datos.tipo_ingreso === "InHouse"
  );
}

export function buscarContratistas(filtro: FiltroContratistas = {}): Promise<PaginaContratistas> {
  return invoke("buscar_contratistas", { filtro });
}

export async function crearContratista(datos: DatosContratista): Promise<number> {
  const id = await invoke<number>("crear_contratista", { datos });
  solicitarSincronizacionNube();
  return id;
}

export async function actualizarContratista(id: number, datos: DatosContratista): Promise<void> {
  await invoke("actualizar_contratista", { id, datos });
  solicitarSincronizacionNube();
}

File: desktop/src-tauri/src/dto/contratistas.rs
use chrono::NaiveDate;

use control_acceso::database::queries::contratistas::FiltroContratistas;
use control_acceso::models::tipo_ingreso::TipoIngreso;
use control_acceso::services::contratista_service::{
    DatosActualizacionContratista, DatosContratista,
};

/// Sólo texto: la grilla de Contratistas ya no manda filtro (carga el
/// universo completo y filtra del lado del cliente con AG Grid — ver
/// `desktop/src/pantallas/Contratistas.tsx`), pero el buscador en vivo de
/// los modales (`NuevoIngresoModal`, `GestionGafeteModal`) sigue
/// necesitando una búsqueda de texto contra el servidor. `limite` siempre
/// pide el máximo (`FiltroContratistas::default()` lo clampea contra
/// `LIMITE_LISTADO_MAXIMO_CARGA_COMPLETA` en
/// `src/database/queries/contratistas.rs`) — ninguno de los dos casos de
/// uso quiere paginar.
#[derive(serde::Deserialize, Default)]
pub struct FiltroContratistasEntrada {
    pub texto: Option<String>,
}

impl FiltroContratistasEntrada {
    pub fn construir(self) -> FiltroContratistas {
        FiltroContratistas {
            texto: self
                .texto
                .map(|t| t.trim().to_owned())
                .filter(|t| !t.is_empty()),
            limite: usize::MAX,
            ..Default::default()
        }
    }
}

/// Espejo de `DatosContratista`/`DatosActualizacionContratista` — el core ya
/// modela crear y editar con dos structs idénticos en forma (distinto nombre
/// nada más), así que del lado del webview alcanza con uno solo. Cubre tanto
/// el formulario completo (crear/editar) como el toggle rápido de
/// "es de ruta"/"tiene acceso" desde la grilla (la fila ya tiene todos los
/// demás campos, se mandan de vuelta sin cambios).
#[derive(serde::Deserialize)]
pub struct DatosContratistaEntrada {
    pub cedula: String,
    pub nombre: String,
    pub empresa_id: i64,
    pub tipo_ingreso: TipoIngreso,
    pub fecha_vencimiento_praind: Option<NaiveDate>,
    pub es_personal_ruta: bool,
    pub tiene_acceso: bool,
}

impl From<DatosContratistaEntrada> for DatosContratista {
    fn from(entrada: DatosContratistaEntrada) -> Self {
        Self {
            cedula: entrada.cedula,
            nombre: entrada.nombre,
            empresa_id: entrada.empresa_id,
            tipo_ingreso: entrada.tipo_ingreso,
            fecha_vencimiento_praind: entrada.fecha_vencimiento_praind,
            es_personal_ruta: entrada.es_personal_ruta,
            tiene_acceso: entrada.tiene_acceso,
        }
    }
}

impl From<DatosContratistaEntrada> for DatosActualizacionContratista {
    fn from(entrada: DatosContratistaEntrada) -> Self {
        Self {
            cedula: entrada.cedula,
            nombre: entrada.nombre,
            empresa_id: entrada.empresa_id,
            tipo_ingreso: entrada.tipo_ingreso,
            fecha_vencimiento_praind: entrada.fecha_vencimiento_praind,
            es_personal_ruta: entrada.es_personal_ruta,
            tiene_acceso: entrada.tiene_acceso,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn el_texto_en_blanco_se_convierte_en_ninguno() {
        let filtro = FiltroContratistasEntrada {
            texto: Some("   ".to_owned()),
        }
        .construir();

        assert_eq!(filtro.texto, None);
    }

    #[test]
    fn el_texto_se_recorta_de_espacios() {
        let filtro = FiltroContratistasEntrada {
            texto: Some("  ana  ".to_owned()),
        }
        .construir();

        assert_eq!(filtro.texto, Some("ana".to_owned()));
    }

    #[test]
    fn sin_texto_pide_el_limite_maximo_sin_paginar() {
        let filtro = FiltroContratistasEntrada::default().construir();

        assert_eq!(filtro.texto, None);
        assert_eq!(filtro.offset, 0);
    }
}
